[ObjCopy][DX] Support for -dump-section flag

llvm-beanz opened the pull request:
This adds support for the `-dump-section=<section>=<file>` flag for the DXContainer file format. This flag dumps the contents of a named section to the specified file.

This flag is particularly handy for ripping DXIL bitcode out of the object files so that we can use LLVM tools to inspect and operate on the bitcode. To facilitate that workflow this flag also strips the program headers from parts containing DXIL so that the resulting file is a valid bitcode file.

llvm-beanz committed:
[ObjCopy][DX] Support for -dump-section flag

This adds support for the -dump-section=<section>=<file> flag for the
DXContainer file format. This flag dumps the contents of a named section
to the specified file.

This flag is particularly handy for ripping DXIL bitcode out of the
object files so that we can use LLVM tools to inspect and operate on
the bitcode. To facilitate that workflow this flag also strips the
program headers from parts containing DXIL so that the resulting file is
a valid bitcode file.
../llvm/lib/ObjCopy/DXContainer/DXContainerObjcopy.cpp
../llvm/test/tools/llvm-objcopy/DXContainer/dump-section.yaml

inbelic reviewed (commented):
Largely LGTM, I will note the pr description seems incomplete/mis-typed

llvm-beanz committed:
Respond to review feedback.
../llvm/lib/ObjCopy/DXContainer/DXContainerObjcopy.cpp
../llvm/test/tools/llvm-objcopy/DXContainer/dump-section-errors.yaml
../llvm/test/tools/llvm-objcopy/DXContainer/dump-section.yaml

llvm-beanz commented on llvm/lib/ObjCopy/DXContainer/DXContainerObjcopy.cpp:
```diff
@@ -42,7 +44,40 @@ static Error extractPartAsObject(StringRef PartName, StringRef OutFilename,
+    if (P.Name == PartName) {
+      ArrayRef<uint8_t> Contents = P.Data;
+      // For parts containing llvm bitcode, don't dump the program headers so
+      // that we get a valid .bc file.
```
I added a comment here to explain better. The gist is that the DXContainer section descriptions are put in at the beginning of the section data as a header. For the bitcode sections, the headers have more data that is just replicated out of the bitcode. This allows runtime tools to avoid cracking open the bitcode to get basic information about of the program.

My goal is to make the objcopy tooling a bit more ergonomic by allowing it to pull out the bitcode into a file that other tools can directly work with. We should also be able to extend the libObject interface to inspect the bitcode and re-materialize the extra header information so that we could support generating a bitcode section and the associated headers from a bitcode file provided as input.

jh7370 replied:
Thanks!

jh7370 commented on llvm/test/tools/llvm-objcopy/DXContainer/dump-section-errors.yaml:
```diff
@@ -0,0 +1,26 @@
+# RUN: not llvm-objcopy --dump-section=FKE3=%t.fek1 %t.dxbc 2>&1 | FileCheck %s --check-prefix=CHECK-MISSING
+# RUN: not llvm-objcopy --dump-section=FKE2=%t/does_not_exist/.fek2 %t.dxbc 2>&1 | FileCheck %s --check-prefix=CHECK-BAD-PATH
+
+# CHECK-EMPTY: error: {{.*}}: part 'FKE0' is empty
```
Let's explicitly match the file name (use `-DFILE=...` in the FileCheck command and `[[FILE]]` here, or equivalent): we've had problems in the past with --dump-section for other formats resulting in the wrong file name being used in error messages.

jh7370 commented on llvm/test/tools/llvm-objcopy/DXContainer/dump-section-errors.yaml:
```diff
@@ -0,0 +1,26 @@
+
+# CHECK-EMPTY: error: {{.*}}: part 'FKE0' is empty
+# CHECK-MISSING: error: {{.*}}: part 'FKE3' not found
+# CHECK-BAD-PATH: error: {{.*}}: No such file or directory
```
This should use the lit substitution for the corresponding error message type, since the message is platform-specific (the test currently fails on Windows). I believe the substitution you'll need is `%errc_ENOENT` (which you should combine with FileCheck's -D option, similar to above).

jh7370 commented on llvm/lib/ObjCopy/DXContainer/DXContainerObjcopy.cpp:
```diff
@@ -42,7 +44,40 @@ static Error extractPartAsObject(StringRef PartName, StringRef OutFilename,
+      Expected<std::unique_ptr<FileOutputBuffer>> BufferOrErr =
+          FileOutputBuffer::create(Filename, Contents.size());
+      if (!BufferOrErr)
+        return createFileError(Filename, BufferOrErr.takeError());
```
I assume it's not practical to test both the `create` and `commit` error paths? (I see one is tested, but not the other).

jh7370 commented on llvm/test/tools/llvm-objcopy/DXContainer/dump-section.yaml:
```diff
@@ -0,0 +1,283 @@
+# RUN: llvm-objcopy --dump-section=PSV0=%t.psv0 %t.dxbc
+# RUN: wc -c %t.psv0 | FileCheck %s --check-prefix=PSV0-SIZE
+# PSV0-SIZE: 76
+--- !dxcontainer
```
Nit: I suggest adding a blank line between the last check and the YAML, to help clarify the YAML isn't tied to that specific case.

jh7370 commented on llvm/test/tools/llvm-objcopy/DXContainer/dump-section.yaml:
```diff
@@ -0,0 +1,283 @@
+      IncludesSource:  false
+      Digest:          [ 0x2, 0xB5, 0x24, 0x1, 0x50, 0x48, 0xC3, 0x12,
+                         0x39, 0x28, 0xAE, 0x46, 0x68, 0x14, 0x84, 0xDA ]
+  - Name:            ISG1
```
Nit: is it possible to drop some of these parts that aren't needed for the test? It'll help future readers see what's important for the test case. (If it's not possible due to the way the data is read, that's fine of course).

jh7370 commented on llvm/test/tools/llvm-objcopy/DXContainer/dump-section.yaml:
```diff
@@ -0,0 +1,283 @@
+## Dump the PSV0 part and verify its size.
+# RUN: llvm-objcopy --dump-section=PSV0=%t.psv0 %t.dxbc
+# RUN: wc -c %t.psv0 | FileCheck %s --check-prefix=PSV0-SIZE
+# PSV0-SIZE: 76
```
If there's a way we can at least partially verify the contents are correct and not just the size, that would be great. I'm slightly worried that this could miss a case where we're writing garbage/all zeros to the file rather than the real data we want.

jh7370 replied:
Adding `--add-section` support seems like a reasonable approach to me. As an interim, you could use something like `od` just to check the first few bytes or something.

llvm-beanz committed:
Fixing up tests based on feedback.
../llvm/test/tools/llvm-objcopy/DXContainer/dump-section-errors.yaml
../llvm/test/tools/llvm-objcopy/DXContainer/dump-section.yaml

jh7370 reviewed (approved):
LGTM. Probably worth waiting for a DXContainer expert to sign off too.

bogner commented on llvm/lib/ObjCopy/DXContainer/DXContainerObjcopy.cpp:
```diff
@@ -42,7 +44,49 @@ static Error extractPartAsObject(StringRef PartName, StringRef OutFilename,
+
 static Error handleArgs(const CommonConfig &Config, Object &Obj) {
+  for (StringRef Flag : Config.DumpSection) {
+    StringRef SecName;
+    StringRef FileName;
+    std::tie(SecName, FileName) = Flag.split("=");
```
We have C++17, you can use a structured binding here if you want.
```suggestion
    auto [SecName, FileName] = Flag.split("=");
```

bogner commented on llvm/test/tools/llvm-objcopy/DXContainer/dump-section-errors.yaml:
```diff
@@ -0,0 +1,27 @@
+# RUN: yaml2obj %s -o %t.dxbc
+# RUN: not llvm-objcopy --dump-section=FKE0=%t.fek0 %t.dxbc 2>&1 | FileCheck %s --check-prefix=CHECK-EMPTY -DFILE=%t.fek0
```
FileCheck has a [CHECK-EMPTY directive](https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-empty-directive) so this prefix could be confusing / ambiguous. Might be worth a different name to avoid that.

jh7370 replied:
As the one who implemented `CHECK-EMPTY`, I should have seen that 😂

bogner reviewed (approved):
LGTM. Just a few take it or leave it nitpicks.

llvm-beanz committed:
Update llvm/lib/ObjCopy/DXContainer/DXContainerObjcopy.cpp

Co-authored-by: Justin Bogner <mail@justinbogner.com>

llvm-beanz committed:
Updates based on review feedback
../llvm/lib/ObjCopy/DXContainer/DXContainerObjcopy.cpp
../llvm/test/tools/llvm-objcopy/DXContainer/dump-section-errors.yaml
